File: web/src/main/java/top/bettercode/summer/web/support/code/CodeService.java
package top.bettercode.summer.web.support.code;

import com.github.benmanes.caffeine.cache.Cache;
import com.github.benmanes.caffeine.cache.Caffeine;
import java.util.concurrent.TimeUnit;
import top.bettercode.summer.tools.lang.property.PropertiesSource;

/**
 * @author Peter Wu
 */
public class CodeService implements ICodeService {

  private final PropertiesSource propertiesSource;

  private final long expireSeconds = 60L;
  private final Cache<String, DicCodes> cache = Caffeine.newBuilder()
      .expireAfterWrite(expireSeconds, TimeUnit.SECONDS).build();

  public CodeService(PropertiesSource propertiesSource) {
    this.propertiesSource = propertiesSource;
  }

  @Override
  public DicCodes getDicCodes(String codeType) {
    return cache.get(codeType,
        type -> new DicCodes(type, propertiesSource.get(type),
            propertiesSource.mapOf(type,
                "Int".equals(propertiesSource.get(type + "|TYPE")))));
  }
}

File: web/src/main/kotlin/top/bettercode/summer/web/support/code/CodeService.kt
package top.bettercode.summer.web.support.code

import com.github.benmanes.caffeine.cache.Caffeine
import top.bettercode.summer.tools.lang.property.PropertiesSource
import java.util.concurrent.TimeUnit

/**
 * @author Peter Wu
 */
class CodeService(private val propertiesSource: PropertiesSource) : ICodeService {
    private val expireSeconds = 60L
    private val cache = Caffeine.newBuilder()
        .expireAfterWrite(expireSeconds, TimeUnit.SECONDS).build<String, DicCodes>()

    val codeTypes: List<String> by lazy {
        propertiesSource.all().keys.filter { !it.endsWith("|TYPE") && !it.contains(".") }
    }

    override fun getDicCodes(codeType: String): DicCodes? {
        return cache[codeType, { type: String ->
            DicCodes(
                type, propertiesSource[type],
                propertiesSource.mapOf(type, propertiesSource["$type|TYPE"])
            )
        }]
    }
}
